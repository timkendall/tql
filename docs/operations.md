# Operations

## Example

```typescript
import { query } from './sdk'

const QUERY = query(t => [
  t.viewer(t => [
    t.id()
  ])
]).toOperation({ name: 'Viewer' })
```