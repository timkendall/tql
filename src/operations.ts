import { TypedDocumentNode } from "@graphql-typed-document-node/core"; // only needed before graphql < 16
import { Concat } from "typescript-tuple";

import {
  Primitive,
  SelectionSet,
  Selection,
  Field,
  Variable,
  Operation,
} from "./Operation";
import { ISelector, selector } from "./Selector";

interface User {
  id: string;
  friends(limit: number): User[];
}

interface Food {
  id: string;
  avgCals(timeframe: string): number;
}

interface Query {
  viewer: User;
  hello(name: bigint, pronoun?: string): string;
  user(id: string): User;
  foods(limit?: number, after?: string): Food[]; // @todo Array<infer T>
}

type ResolverArgs<T> = T extends (...args: any[]) => any ? Parameters<T> : [];

export type Experimental<T> = {
  [F in keyof T]: T[F] extends infer U
    ? U extends (...args: infer Vars) => infer Type // parameterized fields
      ? Type extends Primitive
        ? (
            ...args: Vars
          ) => Field<F extends string ? F : never /*, @todo add arguments */> // dumb
        : <S extends Array<Selection>>(
            ...args: Concat<Vars, [select: (selector: Experimental<Type>) => S]> // @todo add variables
          ) => // select: (selector: Experimental<Type>) => S
          Field<
            F extends string ? F : never,
            never /* @todo add arguments */,
            SelectionSet<S>
          >
      : U extends infer Type // non-paramaterized fields
      ? Type extends Primitive
        ? () => Field<F extends string ? F : never>
        : <S extends Array<Selection>>(
            select: (selector: Experimental<Type>) => S
          ) => Field<F extends string ? F : never, never, SelectionSet<S>>
      : never
    : never;
};

type Test = Experimental<Query>;

const Query = {} as Test;
Query.viewer((t) => [t.id()]);
Query.hello(BigInt(""), "Mr.");
Query.user("name", (t) => [t.id()]);
Query.foods(undefined, undefined, (t) => []);

// // examples: $('id'), $('id!')
// export const $ = (literals: TemplateStringsArray, ...placeholders: string[]) => {
//   return new Variable<typeof literals[0]>(literals[0])
// }

export const $ = <T extends string>(name: T) => {
  return new Variable(name);
};

export function query<T extends Array<Selection>>(
  selections: T
): Operation<SelectionSet<T>>;
export function query<T>(
  select: (selector: ISelector<T>) => Array<Selection>
): Operation<SelectionSet<Array<Selection>>> {
  if (Array.isArray(arguments[0])) {
    return new Operation("", "query", new SelectionSet(arguments[0]));
  } else {
    return new Operation("", "query", new SelectionSet(select(selector())));
  }
}

// query<T, S extends Array<Selection>>(select: (selector: T) => S): TypeDocumentNode<Result<S>, Variables<S>>
// { definitions: ReadonlyArray<ExecutableDefinitionNode> }
