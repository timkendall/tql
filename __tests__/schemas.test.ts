import fs from "fs";
import path from "path";
import { parse, validate, execute, buildSchema } from "graphql";

import * as starwarsOps from "./starwars/operations";
import * as githubOps from "./github/operations";
import * as hasuraOps from "./hasura/operationts";

import { StarWarsSchema } from "./starwars/starwars.schema";

describe("Schemas", () => {
  describe("Star Wars", () => {
    const typeDefs = fs
      .readFileSync(
        path.resolve(__dirname, "./starwars/starwars.schema.graphql")
      )
      .toString("utf-8");
    const schema = buildSchema(typeDefs);

    describe.each([[starwarsOps.kitchensink.name, starwarsOps.kitchensink]])(
      `%s`,
      (_operationName, operation) => {
        it("renders the expected operation", () => {
          expect(operation.toString()).toMatchSnapshot();
        });

        it("renders a parsable operation", () => {
          expect(() => parse(operation.toString())).not.toThrow();
        });

        it("renders a valid operation", () => {
          expect(validate(schema, operation.toDocument()).length).toBe(0);
        });

        it("renders an executable operation", async () => {
          await expect(
            execute({
              schema: StarWarsSchema,
              document: operation.toDocument(),
            })
          ).resolves.toMatchSnapshot();
        });
      }
    );
  });

  describe("GitHub", () => {
    const typeDefs = fs
      .readFileSync(path.resolve(__dirname, "./github/github.schema.graphql"))
      .toString("utf-8");
    const schema = buildSchema(typeDefs);

    describe.each([[githubOps.viewer.name, githubOps.viewer]])(
      `%s`,
      (_operationName, operation) => {
        it("renders the expected operation", () => {
          expect(operation.toString()).toMatchSnapshot();
        });

        it("renders a parsable operation", () => {
          expect(() => parse(operation.toString())).not.toThrow();
        });

        it("renders a valid operation", () => {
          expect(validate(schema, operation.toDocument()).length).toBe(0);
        });
      }
    );
  });

  describe("Hasura", () => {
    const typeDefs = fs
      .readFileSync(path.resolve(__dirname, "./hasura/hasura.schema.graphql"))
      .toString("utf-8");
    const schema = buildSchema(typeDefs);

    describe.each([[hasuraOps.bookmarks.name, hasuraOps.bookmarks]])(
      `%s`,
      (_operationName, operation) => {
        it("renders the expected operation", () => {
          expect(operation.toString()).toMatchSnapshot();
        });

        it("renders a parsable operation", () => {
          expect(() => parse(operation.toString())).not.toThrow();
        });

        it("renders a valid operation", () => {
          expect(validate(schema, operation.toDocument()).length).toBe(0);
        });
      }
    );
  });
});
