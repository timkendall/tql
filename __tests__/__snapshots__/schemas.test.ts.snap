// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Schemas GitHub Viewer renders the expected operation 1`] = `
"query Viewer {
  viewer {
    name
    twitterUsername
    status {
      emoji
      expiresAt
    }
    repositories(first: 3) {
      pageInfo {
        endCursor
        startCursor
        hasNextPage
        hasPreviousPage
      }
      nodes {
        __typename
        id
        name
        createdAt
      }
      edges {
        cursor
        node {
          id
        }
      }
    }
  }
}"
`;

exports[`Schemas Hasura Bookmarks renders the expected operation 1`] = `
"query Bookmarks {
  bookmarks(limit: 10) {
    id
    name
    children_aggregate(distinct_on: createdUtc) {
      aggregate {
        min {
          id
        }
        max {
          id
        }
      }
      nodes {
        id
        createdUtc
      }
    }
  }
  playlist_items_by_pk(id: 1) {
    __typename
    id
    position
  }
}"
`;

exports[`Schemas Star Wars Example renders an executable operation 1`] = `
Object {
  "data": Object {
    "human": Object {
      "__typename": "Human",
      "appearsIn": Array [
        "NEWHOPE",
        "EMPIRE",
        "JEDI",
      ],
      "friends": Array [
        Object {
          "__typename": "Human",
          "appearsIn": Array [
            "NEWHOPE",
            "EMPIRE",
            "JEDI",
          ],
          "homePlanet": "Tatooine",
          "id": "1000",
          "name": "Luke Skywalker",
        },
        Object {
          "__typename": "Human",
          "appearsIn": Array [
            "NEWHOPE",
            "EMPIRE",
            "JEDI",
          ],
          "homePlanet": "Alderaan",
          "id": "1003",
          "name": "Leia Organa",
        },
        Object {
          "__typename": "Droid",
          "appearsIn": Array [
            "NEWHOPE",
            "EMPIRE",
            "JEDI",
          ],
          "id": "2001",
          "name": "R2-D2",
          "primaryFunction": "Astromech",
        },
      ],
      "homePlanet": null,
      "id": "1002",
      "mass": 80,
      "name": "Han Solo",
      "starships": Array [
        Object {
          "id": "3000",
          "name": "Millenium Falcon",
        },
        Object {
          "id": "3003",
          "name": "Imperial shuttle",
        },
      ],
    },
    "reviews": Array [],
    "search": Array [
      Object {
        "__typename": "Human",
        "id": "1002",
        "name": "Han Solo",
      },
    ],
  },
}
`;

exports[`Schemas Star Wars Example renders the expected operation 1`] = `
"query Example {
  search(text: \\"han\\") {
    __typename
    ... on Human {
      id
      name
    }
  }
  reviews(episode: EMPIRE) {
    stars
    commentary
  }
  human(id: \\"1002\\") {
    __typename
    id
    name
    appearsIn
    homePlanet
    mass
    friends {
      __typename
      id
      name
      appearsIn
      ... on Human {
        homePlanet
      }
      ... on Droid {
        primaryFunction
      }
    }
    starships {
      id
      name
    }
  }
}"
`;
